src/aof.c

```diff
@@ -40,189 +40,692 @@
+void aofInfoFree(aofInfo *ai) {
+    serverAssert(ai != NULL);
+    if (ai->file_name) sdsfree(ai->file_name);
+    zfree(ai);
+}
+
+/* Deep copy an aofInfo. */
+aofInfo *aofInfoDup(aofInfo *orig) {
+    serverAssert(orig != NULL);
+    aofInfo *ai = aofInfoCreate();
+    ai->file_name = sdsdup(orig->file_name);
+    ai->file_seq = orig->file_seq;
+    ai->file_type = orig->file_type;
+    return ai;
+}
+
+/* Method to free AOF list elements. */
+void aofListFree(void *item) {
+    aofInfo *ai = (aofInfo *)item;
+    aofInfoFree(ai);
+}
+
+/* Method to duplicate AOF list elements. */
+void *aofListDup(void *item) {
+    return aofInfoDup(item);
 }
 
-/* Return the current size of the AOF rewrite buffer. */
-unsigned long aofRewriteBufferSize(void) {
+/* Create an empty aofManifest, which will be called in `aofLoadManifestFromDisk`. */
+aofManifest *aofManifestCreate(void) {
+    aofManifest *am = zcalloc(sizeof(aofManifest));
+    am->incr_aof_list = listCreate();
+    am->history_aof_list = listCreate();
+    listSetFreeMethod(am->incr_aof_list, aofListFree);
+    listSetDupMethod(am->incr_aof_list, aofListDup);
+    listSetFreeMethod(am->history_aof_list, aofListFree);
+    listSetDupMethod(am->history_aof_list, aofListDup);
+    return am;
+}
+
+/* Free the aofManifest structure (pointed to by am) and its embedded members. */
+void aofManifestFree(aofManifest *am) {
+    if (am->base_aof_info) aofInfoFree(am->base_aof_info);
+    if (am->incr_aof_list) listRelease(am->incr_aof_list);
+    if (am->history_aof_list) listRelease(am->history_aof_list);
+    zfree(am);
+}
+
+sds getAofManifestFileName() {
+    return sdscatprintf(sdsempty(), "%s%s", server.aof_filename, 
+                MANIFEST_NAME_SUFFIX);
+}
+
+sds getTempAofManifestFileName() {
+    return sdscatprintf(sdsempty(), "%s%s%s", MANIFEST_TEMP_NAME_PREFIX, 
+                server.aof_filename, MANIFEST_NAME_SUFFIX);
+}
+
+/* Returns the string representation of aofManifest pointed to by am.
+ * 
+ * The string is multiple lines separated by '\n', and each line represents
+ * an AOF file.
+ * 
+ * Each line contains 6 fields (they are separated by spaces). Among 
+ * them, the 0th, 2nd, and 4th field respectively represent the manifest 
+ * key, they are:
+ * file: AOF file name
+ * seq: The serial number of the AOF file
+ * type: Types of AOF file, There are: b(BASE)、h(HISTORY)、i(INCR)
+ * 
+ * A possible line:
+ *    file appendonly_1.base.rdb seq 1 type b
+ * 
+ * The BASE file information (if we have) will be placed on the first 
+ * line, followed by history type AOFs and finally is the INCR AOFs.
+ */
+#define AOF_INFO_FORMAT_AND_CAT(buf, info)                    \
+    sdscatprintf(buf, "%s %s %s %lld %s %c\n",                \
+                 AOF_MANIFEST_KEY_FILE_NAME, info->file_name, \
+                 AOF_MANIFEST_KEY_FILE_SEQ, info->file_seq,   \
+                 AOF_MANIFEST_KEY_FILE_TYPE, info->file_type)
```

Comment:
```suggestion
    sdscatprintf((buf), "%s %s %s %lld %s %c\n",                \
                 AOF_MANIFEST_KEY_FILE_NAME, (info)->file_name, \
                 AOF_MANIFEST_KEY_FILE_SEQ, (info)->file_seq,   \
                 AOF_MANIFEST_KEY_FILE_TYPE, (info)->file_type)
```

Safer use of macro arguments. BTW do we really need a macro here?

Reply:
I think it is necessary. Although there is only one `sdscatprintf` function in the macro, the definition of the format is repeated. Imagine that we need to change the format of the manifest in the future (for example, add new fields in the back), then we only need to change the macro.